### ISSUE TITLE
Feature: hooks for image loads / uploads for URL preprocessing / postprocessing

### ISSUE THREAD

**Issue by vadimkantorov** [facebook/lexical#6255]
- In markdown documents we often have relative URLs to images which should be customly mapped to some download_url before display. On the other direction as well, it would be nice to support a hook for uploading images to a directory (with a custom code) and replacing the base64 datauri with an absolute or relative URL to the newly uploaded image (base directory would often be the images directory). For uploads this functionality would be similar to how we upload images in GitHub issue comments: the file is uploaded to the server and then the URL is used (except in GitHub the absolute URLs are used, whereas in a small blog/website context, being able to use relative URLs along with a base directory for images is more natural)

- Also, for Markdown editor, when editing source code even when images are embedded as datauris, it would be nice to allow to replace the datauris with some special macros/values (e.g. `![](lexical://tmpimage1.jpg)` ) to facilitate manual editing of Markdown using datauri's for image embeds. I guess, this can be somehow implemented in the Markdown plugin, or even in user code, but an example of achieving this would increase the practicality of the Markdown editing mode when working with images

I also found these bugs when trying out image pasting functionality in the Playground: 
- https://github.com/facebook/lexical/issues/6256
- https://github.com/facebook/lexical/issues/5474

> Context: I'm building a simplistic single-file 500LOC analog of https://mdxeditor.dev/ for RTF editing of Markdown files stored in GitHub repos, manipulated with GitHub API, supporting uploading/displaying images (available at https://vadimkantorov.github.io/moncms).

**Comment by StyleT** [comment]
Housekeeping is here :P Closed in favor of https://github.com/facebook/lexical/issues/6256

**Comment by vadimkantorov** [comment]
So the way I solved is:
- Introducing ImageCache class with `resolve` method (and also `put` method and `prefix` method)
- Introduce `ImageCacheContext` (with `createContext`)
- In ImageComponent: doing `useContext` and invoking `resolve` just before emitting the img tag
- Wrapping the LexicalComposer in `ImageCacheContext.Provider`

I think it's quite a nice solution and would allow for any flexible image path scheme (e.g. in my context it can take some minutes before the uploaded image is available on the prod server - GitHub Actions static site generation is not instant and GitHub Actions can be down, so being able to use some different image path prefix in editing time is great)

In ImageNode I introduced public `setSrc` which allows me to update the image url once image upload is committed to the server.

In image picker dialogue I added using `createObjectURL` and returning it as `src`, and dumping `datauri` into a separate field. Blob URLs work great as temporary image URLs when working editing Markdown code directly, and it also bypasses the syntax highlighting slowness of https://github.com/facebook/lexical/issues/6256

Overall, I think the `ImageNode` functionality (at least the rendering part) deserves to be promoted out of Playground to core or a separate package.

---

A good test for core nodes can be: importing and displaying all blocks from Notion API. I think now Lexical already has most of them in core or in Playground: https://developers.notion.com/reference/block, so promoting the mature ones out of Playground (such as images, YouTube embeds, emojis) can be nice. I built a Notion exported before: https://github.com/vadimkantorov/notionexportstatic, dumping the JSON from a Notion document via API is quite easy.
